File: redfox-app/src/main/kotlin/io/github/malczuuu/redfox/app/core/UserService.kt
package io.github.malczuuu.redfox.app.core

import io.github.malczuuu.redfox.app.common.Identity
import io.github.malczuuu.redfox.app.common.PageResult
import io.github.malczuuu.redfox.app.domain.UserEntity
import io.github.malczuuu.redfox.app.domain.UserRepository
import java.time.Clock
import java.time.Instant
import java.util.UUID
import org.springframework.data.domain.Pageable
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val clock: Clock,
) {

  fun getUsers(pageable: Pageable): PageResult<UserDto> {
    val users = userRepository.findAll(pageable)
    return PageResult(
        content = users.content.map { it.toDto() },
        page = users.number,
        size = users.size,
        totalElements = users.totalElements,
    )
  }

  fun getUser(id: UUID): UserDto {
    val user = userRepository.findOne(id).orElseThrow { UserNotFoundException() }
    return user.toDto()
  }

  @Transactional
  fun createUser(request: CreateUserDto): Identity {
    if (userRepository.existsByLogin(request.login)) {
      throw UserAlreadyExistsException()
    }
    var user =
        UserEntity(
            login = request.login,
            passhash = passwordEncoder.encode(request.passhash)!!,
            firstName = request.firstName,
            lastName = request.lastName,
        )
    user = userRepository.save(user)
    return Identity(user.id.toString())
  }

  @Transactional
  fun updateUser(id: UUID, request: UpdateUserDto) {
    val user = userRepository.lockOne(id).orElseThrow { UserNotFoundException() }
    user.firstName = request.firstName
    user.lastName = request.lastName
    userRepository.save(user)
  }

  @Transactional
  fun deleteUser(id: UUID) {
    userRepository.lockOne(id).ifPresent { user ->
      user.deletedAt = Instant.now(clock)
      userRepository.save(user)
    }
  }

  private fun UserEntity.toDto() =
      UserDto(
          id = id,
          login = login,
          firstName = firstName,
          lastName = lastName,
          createdAt = createdAt,
          updatedAt = updatedAt,
          version = version,
      )
}

File: redfox-webapp/src/app/core/user/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Identity, PageResult } from '../common/common.model';
import { CreateUser, UpdateUser, User } from './user.model';

@Injectable({ providedIn: 'root' })
export class UserService {
  private readonly baseUrl = '/api/v1/users';

  constructor(private readonly http: HttpClient) {}

  getUsers(page = 0, size = 20): Observable<PageResult<User>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageResult<User>>(this.baseUrl, { params });
  }

  getUser(id: string): Observable<User> {
    return this.http.get<User>(`${this.baseUrl}/${id}`);
  }

  createUser(request: CreateUser): Observable<Identity> {
    return this.http.post<Identity>(this.baseUrl, request);
  }

  updateUser(id: string, request: UpdateUser): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}`, request);
  }

  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
